frontend/ui/widget/menu.lua

```diff
@@ -966,13 +966,16 @@ function Menu:init()
     end
 
     if Device:hasDPad() then
-        -- we won't catch presses to "Right", leave that to MenuItem.
-        self.key_events.FocusRight = nil
+        if not Device:hasFiveWay() then
+            -- we won't catch presses to "Right", leave that to MenuItem.
+            self.key_events.FocusRight = nil
+            -- add long press on "Right" key
+            self.key_events.Right = { { "Right" } }
+        end
```

Comment:
Yes, it works great as long as you use list displays. Maybe make it not hasKeyboard and not hasFiveWay.

Reply:
But that would exclude kindles with keyboards where there is also a [two actually] global long press... unless i am missing something, Surely it is fine the way it is

Reply:
> unless i am missing something

That I meant `or` and that your own argument says so too. ;-)

Reply:
Still don't quite see the purpose of the not `hasKeyboard`, not `hasFiveWay` means goodbye all non touch kindles...

Reply:
The purpose is that Kindles are Kindles and not Kindles aren't Kindles.

> Furthermore it would mean PB doesn't get long press on Right...

The other way round then... the point is either it's desired or it's not.

Reply:
`not hasKeyboard and not hasFiveWay` does not work as it applies long press on Right to kindles with keyboards. bad...

`not hasKeyboard or not hasFiveWay` also does not work as it then does not apply long press on Right to devices like PB that do need it and currently support it... bad again. Unless this is what you want, in which case, why do you want to remove that partial long press from PB?

the whole idea behind moving long press to `Shift` or `ScreenKB` + `Press` is that it is global and we can remove it from `Right`, independently of grid or list view. 

So why do you want it? what is wrong with just `not hasFiveWay`? I genuinely don't understand your point, eli5 please.

Reply:
`hasKeyboard or not hasFiveway`

> So why do you want it? what is wrong with just `not hasFiveWay`? I genuinely don't understand your point, eli5 please.

Either right to hold is a desirable feature that should be everywhere or it's not and it shouldn't be anywhere except where it's absolutely necessary as a workaround. `hasFiveWay` is being used here to mean something like `canTriggerHoldByKeys` which is a problem because it actually means `isKindle`. I have no idea why @NiLuJe supported the concept because it's code smell plain and simple.

Reply:
`hasKeyboard or not hasFiveway` still keeps right on kindles with keyboards.

`hasKeyboard and not hasFiveway` would add to keyboard emulator but not to PB.

> Either right to hold is a desirable feature that should be everywhere or it's not and it shouldn't be anywhere except where it's absolutely necessary as a workaround

We are in agreement here, but having the `hasKeyboard` doesn't seem to help as far as i can see.

The rest, don't shoot the messenger!

Reply:
To summarize, imho when reading the text `hasFiveWay` it implies one thing and one thing only: that you can click/press on the d-pad or stick itself as a "fifth" action.[1] This is almost entirely useless information, except potentially to see it as a kind of restriction: maybe it's difficult to press on it without accidentally triggering a direction. For all intents and purposes it simply means we have four directions and an a primary action key.[2]

Instead it's being used as a secret code for two or three other features: that page turning buttons exist, and that some sufficient selection of other additional buttons exist so that actions like hold can be triggered.

Basically what you just witnessed above is me being confused by the opaque terminology already. And apparently it's actively getting in the way of correct behavior to boot. :-/

[1] Click is the more commonly accepted vocabulary because press or push could more easily be misinterpreted (as in you press or push it in a certain direction).

[2] There could also be a "true" primary action button, leaving the "fifth" click action as a potentially hidden additional button. But in that case we'd either ignore it or map it to an action like `Hold` directly, so there seems to be no need to distinguish that.

Reply:
> Basically what you just witnessed above is me being confused by the opaque terminology already.

We all have days like that. No worries 😉

Reply:
> What do you mean?

It now has a different behavior on right everywhere but Kindle.

Reply:
Well, if a kind soul could fix the issue with the grid. Potentially the emu and anything with keyboards could have Right as right and something else as hold. Just saying... but we can only talk about that when and if that happens

Reply:
It's my mistake merging this, but I should've rejected the change.

Reply:
And everything else already has something else as hold. What it has now is a weird inconsistency which makes maintenance harder.

Reply:
Okay, let’s take a breath and think. How would you like to approach this? Should we use hasFewKeys instead then? That way it continues to be the odd one out.

Reply:
The behavior should in all likelihood indeed exist on hasFewKeys, but what's lacking from the perspective of this here is something like canTriggerHold (ick :-). But in any case it's more of an inconvenience and less of a "why does this device have so few keys".